libs/mrwebrtc/src/media/device_video_track_source.cpp

```diff
@@ -252,29 +296,449 @@ ErrorOr<RefPtr<DeviceVideoTrackSource>> DeviceVideoTrackSource::Create(
+      webrtc::GetClass(
+          env, "com/microsoft/mixedreality/webrtc/AndroidCameraInterop");
+  RTC_DCHECK(android_camera_interop_class.obj() != nullptr)
+      << "Failed to find AndroidCameraInterop Java class.";
+  jmethodID create_texture_helper_method = webrtc::GetStaticMethodID(
+      env, android_camera_interop_class.obj(), "CreateSurfaceTextureHelper",
+      "()Lorg/webrtc/SurfaceTextureHelper;");
+  jobject texture_helper = env->CallStaticObjectMethod(
+      android_camera_interop_class.obj(), create_texture_helper_method);
+  CHECK_EXCEPTION(env);
+  RTC_DCHECK(texture_helper != nullptr)
+      << "Cannot get the Surface Texture Helper.";
+
+  // Create the video track source which wraps the Android camera capturer
+  rtc::scoped_refptr<webrtc::jni::AndroidVideoTrackSource> impl_source(
+      new rtc::RefCountedObject<webrtc::jni::AndroidVideoTrackSource>(
+          global_factory->GetSignalingThread(), env, false));
+  rtc::scoped_refptr<webrtc::VideoTrackSourceProxy> proxy_source =
+      webrtc::VideoTrackSourceProxy::Create(
+          global_factory->GetSignalingThread(),
+          global_factory->GetWorkerThread(), impl_source);
+
+  int width = 0;
+  if (init_config.width > 0) {
+    width = (int)init_config.width;
+  }
+  int height = 0;
+  if (init_config.height > 0) {
+    height = (int)init_config.height;
+  }
+  float framerate = 0.0f;
+  if (init_config.framerate > 0.0f) {
+    framerate = (float)init_config.framerate;
+  }
+
+  // Create the camera capturer and bind it to the surface texture and the video
+  // source, then start capturing.
+  jmethodID start_capture_method = webrtc::GetStaticMethodID(
+      env, android_camera_interop_class.obj(), "StartCapture",
+      "(JLorg/webrtc/SurfaceTextureHelper;Ljava/lang/String;IIF)Lorg/webrtc/VideoCapturer;");
+  jstring java_device_name = env->NewStringUTF(init_config.video_device_id);
+  CHECK_EXCEPTION(env);
+  jobject camera_tmp = env->CallStaticObjectMethod(
+      android_camera_interop_class.obj(), start_capture_method, (jlong)proxy_source.get(),
+      texture_helper, java_device_name, (jint)width, (jint)height, (jfloat)framerate);
+  CHECK_EXCEPTION(env);
+
+  // Java objects created are always returned as local references; create a new
+  // global reference to keep the camera capturer alive.
+  jobject java_video_capturer = (jobject)env->NewGlobalRef(camera_tmp);
+
+  video_source = proxy_source;
+
+#endif  // !defined(MR_SHARING_ANDROID)
+
   if (!video_source) {
+    RTC_LOG(LS_ERROR) << "Failed to create video track source.";
     return Error(Result::kUnknownError);
   }
 
   // Create the wrapper
   RefPtr<DeviceVideoTrackSource> wrapper =
-      new DeviceVideoTrackSource(global_factory, std::move(video_source));
+      new DeviceVideoTrackSource(global_factory, std::move(video_source)
+#if defined(MR_SHARING_ANDROID)
+                                                     ,
+                                 java_video_capturer
```

Comment:
nit: weird formatting

Reply:
**Coding style rules**
Rule 1/1 : `clang-format` is always right.